File: backend/app/services/clinic_roles.py
"""Роли клиники: Куратор, Администратор, Бухгалтер — матрица доступов."""

from __future__ import annotations

from app.models import UserRole
from app.services.booking_directions import is_course_like_direction_name

# Чаты + канбан + онлайн-запись + KPI
ADMINISTRATOR_ROLES = frozenset({UserRole.administrator})

# Онлайн-запись + дебиторка только курсы/протоколы
CURATOR_ROLES = frozenset({UserRole.curator})

# Финансы (отчёт компании) + расходы
ACCOUNTANT_ROLES = frozenset({UserRole.accountant})


def is_administrator(role: UserRole) -> bool:
    return role == UserRole.administrator


def is_curator(role: UserRole) -> bool:
    return role == UserRole.curator


def can_access_booking(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.admin,
        UserRole.administrator,
        UserRole.manager,
        UserRole.expert,
        UserRole.curator,
    )


def can_access_kpi(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.admin,
        UserRole.administrator,
        UserRole.manager,
    )


def can_access_debtors(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.admin,
        UserRole.administrator,
        UserRole.curator,
    )


def can_access_company_report(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.accountant,
    )


def can_access_expenses(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.accountant,
        UserRole.admin,
    )


def can_access_kanban(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.admin,
        UserRole.administrator,
        UserRole.expert,
    )


def can_access_all_chats(role: UserRole) -> bool:
    return role in (
        UserRole.owner,
        UserRole.super_owner,
        UserRole.admin,
        UserRole.administrator,
        UserRole.manager,
    )


def debtors_course_protocol_only(role: UserRole) -> bool:
    """Куратор видит только долги по курсам и протоколам."""
    return role == UserRole.curator


def is_course_or_protocol_indicator(name: str | None) -> bool:
    return is_course_like_direction_name(name)

File: frontend/src/lib/clinicRoles.ts
/** Клиентские хелперы ролей Куратор / Администратор / Бухгалтер. */

import type { UserRole } from "@/lib/types";

export function roleLabelRu(role: UserRole | string | null | undefined): string {
  switch (role) {
    case "owner":
      return "Владелец";
    case "manager":
      return "Менеджер";
    case "expert":
      return "Эксперт";
    case "admin":
      return "Админ воронки";
    case "administrator":
      return "Администратор";
    case "curator":
      return "Куратор";
    case "accountant":
      return "Бухгалтер";
    case "finance_analyst":
      return "Фин. аналитик";
    case "super_owner":
      return "Супер-владелец";
    case "rop":
      return "РОП";
    default:
      return role || "—";
  }
}

export function isAdministrator(role: UserRole | null): boolean {
  return role === "administrator";
}

export function isCurator(role: UserRole | null): boolean {
  return role === "curator";
}

export function isAccountant(role: UserRole | null): boolean {
  return role === "accountant";
}

export function isRop(role: UserRole | null): boolean {
  return role === "rop";
}

export function canAccessRop(role: UserRole | null): boolean {
  return role === "rop" || role === "owner" || role === "super_owner";
}

export function canAccessCuratorJournal(role: UserRole | null): boolean {
  return (
    role === "owner" ||
    role === "super_owner" ||
    role === "admin" ||
    role === "administrator" ||
    role === "curator"
  );
}
